### PR TITLE
Enable deterministic Cosmwasm stargate queries

### PR DESCRIPTION
**Pull request by mattverse**
## What is the purpose of the change
cref: https://github.com/osmosis-labs/wasmd/pull/7

The original design was to have basic initialization for the whitelisted quries in the wasmd side, but upon leaving the inititialized whitelist empty, this is the attempt to move the whitelisted quries and bindings layer into osmosis side. 

## Brief Changelog

- Add whitelisted stargate queries and ensure deterministic query.

## Testing and Verifying
Further plans to test this is TBD(upon concept ACK of this approach), although I do not have exact plans to test this in a local machine as of rn.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: wasmbinding/stargate_bindings.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// StargateLayerBindings keeps whitelist and its deterministic
+// response binding for stargate queries.
+//
+// The query can be multi-thread, so we have to use
+// thread safe sync.Map instead map[string]bool.
+var StargateLayerBindings sync.Map
+
+func init() {
+	// auth
+	StargateLayerBindings.Store("/cosmos.auth.v1beta1.Query/Account", &auth.QueryAccountResponse{})
```
Comment (mattverse) [inline wasmbinding/stargate_bindings.go:24]: I'm wondering if we can use these protos directly from `github.com/cosmos/cosmos-sdk` packages instead of defining them within our repo

File: wasmbinding/stargate_bindings.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/osmosis-labs/osmosis/v10/wasmbinding/stargate/mint"
+	"github.com/osmosis-labs/osmosis/v10/wasmbinding/stargate/slashing"
+	"github.com/osmosis-labs/osmosis/v10/wasmbinding/stargate/staking"
+)
+
+// StargateLayerBindings keeps whitelist and its deterministic
+// response binding for stargate queries.
+//
+// The query can be multi-thread, so we have to use
+// thread safe sync.Map instead map[string]bool.
```
Comment (p0mvn) [inline wasmbinding/stargate_bindings.go:19]: It seems that the value in this map is not `bool` but the binding itself. If that's the case, can we avoid mentioning `map[string]bool` or change the value, please? I got stumbled by this for a sec

File: wasmbinding/stargate_bindings.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+var StargateLayerBindings sync.Map
```
Comment (p0mvn) [inline wasmbinding/stargate_bindings.go:20]: Suggest the name to contain "whitelist" in it.

File: wasmbinding/stargate_bindings.go (hunk 1 above)
Comment (p0mvn) [reply wasmbinding/stargate_bindings.go:24]: I think we should be using them directly from the sdk. 

In https://github.com/osmosis-labs/wasmd/pull/7/files, I noticed that there is a function called "NormalizeResponses".

IMO, there is no need for it if we are importing these from the SDK

File: wasmbinding/wasm.go [hunk 4]
```diff
@@ -28,3 +31,13 @@ func RegisterCustomPlugins(
 		messengerDecoratorOpt,
 	}
 }
+
+func RegisterStargateQueries(queryRouter baseapp.GRPCQueryRouter) []wasmkeeper.Option {
```
Comment (p0mvn) [inline wasmbinding/wasm.go:35]: How are you planning to propagate `StargateLayerBindings` to the logic in `wasmd`

I'm assuming the plan is to remove the `wasmd` copy of `StargateLayerBindings` and use this instead?

File: wasmbinding/stargate_bindings.go (hunk 2 above)
Comment (mattverse) [reply wasmbinding/stargate_bindings.go:11]: Ah my bad, fixed!

File: wasmbinding/wasm.go (hunk 4 above)
Comment (mattverse) [reply wasmbinding/wasm.go:35]: We are basically adding the newly added stargatelayerbindings rather than overriding it. This is possible because the stargate whitelist is initialized empty

https://github.com/osmosis-labs/osmosis/pull/2190/files#diff-1ac782d2fab1d48ad7f7ef06cf2d58b37c99b6b4c98879a5ad86a985b73fad58R321

**Review by alexanderbez — COMMENTED** [review]
What's the conclusion we came to on today's call re stargate queries? I'd like to get clarification prior to reviewing this :)

**Comment by mattverse** [comment]
@alexanderbez For twap, we're planning to use bindings. However, we also do want to work towards enabling whitelisted stargate queries, this PR should be good for a round of review!

File: wasmbinding/query_plugin.go [hunk 5]
```diff
@@ -5,12 +5,46 @@ import (
...
+	"github.com/cosmos/cosmos-sdk/baseapp"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	proto "github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/osmosis-labs/osmosis/v10/wasmbinding/bindings"
 )
 
+func StarGateQuerier(queryRouter baseapp.GRPCQueryRouter) func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
```
Comment (alexanderbez) [inline wasmbinding/query_plugin.go:17]: nit: also add a godoc :)

```suggestion
func StargateQuerier(queryRouter baseapp.GRPCQueryRouter) func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
```

File: wasmbinding/query_plugin.go [hunk 6]
```diff
@@ -5,12 +5,46 @@ import (
...
+		route := queryRouter.Route(request.Path)
+		if route == nil {
+			return nil, wasmvmtypes.UnsupportedRequest{Kind: fmt.Sprintf("No route to query '%s'", request.Path)}
+		}
+
+		res, err := route(ctx, abci.RequestQuery{
+			Data: request.Data,
+			Path: request.Path,
+		})
+
```
Comment (alexanderbez) [inline wasmbinding/query_plugin.go:33]:
```suggestion
```

File: wasmbinding/query_plugin.go [hunk 7]
```diff
@@ -110,6 +144,31 @@ func CustomQuerier(qp *QueryPlugin) func(ctx sdk.Context, request json.RawMessag
 	}
 }
 
+func NormalizeReponses(binding interface{}, bz []byte) ([]byte, error) {
```
Comment (alexanderbez) [inline wasmbinding/query_plugin.go:147]: nit: godoc

File: wasmbinding/query_plugin.go [hunk 8]
```diff
@@ -110,6 +144,31 @@ func CustomQuerier(qp *QueryPlugin) func(ctx sdk.Context, request json.RawMessag
...
+	}
+
+	// build new deterministic response
+	bz, err = proto.Marshal(message)
+	if err != nil {
+		return nil, wasmvmtypes.Unknown{}
+	}
+
+	// clear proto message
+	message.Reset()
```
Comment (alexanderbez) [inline wasmbinding/query_plugin.go:167]: why is this necessary?

**Review by alexanderbez — COMMENTED** [review]
LGTM

File: wasmbinding/query_plugin.go (hunk 8 above)
Comment (mattverse) [reply wasmbinding/query_plugin.go:167]: @alexanderbez  tbh, I do not have full understanding why this is here left it here for now as I wanted to be careful of removing what I don't know it does. It was one of the chunks that I've forked from other repository that was using stargate query after enabling it. From my brief understanding, it seems that we are doing this because it was an Any value, and it would cached the old value. 

Looking at it again, I do feel that it might be safe to remove this line since we're not using the `message` varaible any more. Can you give opinions for this?

Comment (alexanderbez) [reply wasmbinding/query_plugin.go:172]: mehh, honestly it's fine to leave it. I was just curious. I asked because `message` is created in the function and never leaves it.

**Review by alexanderbez — APPROVED** [review]

File: wasmbinding/query_plugin.go [hunk 9]
```diff
@@ -5,12 +5,46 @@ import (
...
+	proto "github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/osmosis-labs/osmosis/v10/wasmbinding/bindings"
 )
 
+// StargateQuerier dispatches whitelisted stargate queries
+func StargateQuerier(queryRouter baseapp.GRPCQueryRouter) func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
+	return func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
+		binding, whitelisted := StargateWhitelist.Load(request.Path)
```
Comment (ValarDragon) [inline wasmbinding/query_plugin.go:20]: Just want to be extra-sure, this path is the `/cosmos.auth.v1beta1.Query/Account` path, NOT the http path in the proto files right?

Comment (mattverse) [reply wasmbinding/query_plugin.go:20]: yes, they are defined in the whitelist file

File: wasmbinding/query_plugin.go [hunk 10]
```diff
@@ -5,12 +5,46 @@ import (
...
+		res, err := route(ctx, abci.RequestQuery{
+			Data: request.Data,
+			Path: request.Path,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// normalize response to ensure backward compatibility
+		bz, err := NormalizeReponse(binding, res.Value)
```
Comment (p0mvn) [inline wasmbinding/query_plugin.go:39]:
```suggestion
		bz, err := NormalizeResponse(binding, res.Value)
```

File: wasmbinding/query_plugin.go [hunk 11]
```diff
@@ -110,6 +144,33 @@ func CustomQuerier(qp *QueryPlugin) func(ctx sdk.Context, request json.RawMessag
 	}
 }
 
+// NormalizeReponses normalizes the responses by unmarshalling the response then marshalling them again.
+// Normalizing the response is specifically important for responses that contain type of Any.
+func NormalizeReponse(binding interface{}, bz []byte) ([]byte, error) {
```
Comment (p0mvn) [inline wasmbinding/query_plugin.go:149]: Why is this normalization needed? Since we are not defining custom query protos in our repo and, instead, are directly using what is in the sdk and/or Osmosis, what are we ensuring backward compatibility against?

**Comment by mattverse** [comment]
Found out that we need to Json Marshal the responses because cosmwasm stargate querier json unmarshals the responses by default. Closing this PR and opening a new PR replacing this one
